File: vi/codes/cpp/chapter_stack_and_queue/deque.cpp
/**
 * File: deque.cpp
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    /* Khởi tạo hàng đợi hai đầu */
    deque<int> deque;

    /* Thêm phần tử vào hàng đợi */
    deque.push_back(2);
    deque.push_back(5);
    deque.push_back(4);
    deque.push_front(3);
    deque.push_front(1);
    cout << "Double-ended queue deque = ";
    printDeque(deque);

    /* Truy cập phần tử */
    int front = deque.front();
    cout << "Front element front = " << front << endl;
    int back = deque.back();
    cout << "Back element back = " << back << endl;

    /* Xóa phần tử khỏi hàng đợi */
    deque.pop_front();
    cout << "Front dequeue element popFront = " << front << ", after front dequeue, deque = ";
    printDeque(deque);
    deque.pop_back();
    cout << "Rear dequeue element popLast = " << back << ", after rear dequeue, deque = ";
    printDeque(deque);

    /* Lấy độ dài của hàng đợi hai đầu */
    int size = deque.size();
    cout << "Double-ended queue length size = " << size << endl;

    /* Kiểm tra hàng đợi hai đầu có rỗng không */
    bool empty = deque.empty();
    cout << "Double-ended queue is empty = " << empty << endl;

    return 0;
}

File: vi/codes/python/chapter_stack_and_queue/deque.py
"""
File: deque.py
Created Time: 2022-11-29
Author: Peng Chen (pengchzn@gmail.com)
"""

from collections import deque

"""Code minh họa"""
if __name__ == "__main__":
    # Khởi tạo hàng đợi hai đầu
    deq: deque[int] = deque()

    # Thêm các phần tử vào hàng đợi
    deq.append(2)  # Thêm vào cuối hàng đợi
    deq.append(5)
    deq.append(4)
    deq.appendleft(3)  # Thêm vào đầu hàng đợi
    deq.appendleft(1)
    print("double-ended queue deque =", deq)

    # Truy cập các phần tử
    front: int = deq[0]  # Phần tử ở đầu hàng đợi
    print("Front of the queue element front =", front)
    rear: int = deq[-1]  # Phần tử ở cuối hàng đợi
    print("Rear of the queue element rear =", rear)

    # Xóa các phần tử khỏi hàng đợi
    pop_front: int = deq.popleft()  # Xóa phần tử ở đầu hàng đợi
    print("Front dequeued element  pop_front =", pop_front)
    print("After front dequeue deque =", deq)
    pop_rear: int = deq.pop()  # Xóa phần tử ở cuối hàng đợi
    print("Rear dequeued element  pop_rear =", pop_rear)
    print("After rear dequeue deque =", deq)

    # Lấy độ dài của hàng đợi hai đầu
    size: int = len(deq)
    print("Length of the double-ended queue size =", size)

    # Kiểm tra hàng đợi hai đầu có rỗng hay không
    is_empty: bool = len(deq) == 0
    print("Is the double-ended queue empty =", is_empty)
